module/zfs/arc.c

```diff
@@ -5706,6 +5754,12 @@ arc_read(zio_t *pio, spa_t *spa, const blkptr_t *bp,
 			rc = SET_ERROR(ECKSUM);
 			goto out;
 		}
+		if (cached_only) {
+			if (hdr)
+				mutex_exit(hash_lock);
+			done(NULL, zb, bp, NULL, private);
+			return (0);
+		}
```

Comment:
Similar to the comment above, I'm not sure why we always return 0 here regardless of if `hdr == NULL`.  Also, what's intended behavior for embedded block pointers?

Reply:
What do you think would be the correct behavior there?

Reply:
Based solely on the description of the goal of this PR I'd expect it to only return 0 when a cached version of the required block was found.  For an embedded BP we should be able to always return that data.  Otherwise, return `ENOENT`.

Given that nothing currently uses this code in the PR I think it would be best to include these changes in the larger async CoW PR.  That will make the change larger, but it's difficult to review this portion of the proposed work without that additional context.